### PR TITLE
[Feature] Support fp8 e5m2 kv cache with flashinfer

### PR DESCRIPTION
**Pull request by ispobock**
## Motivation

Support fp8 e5m2 kv cache with flashinfer.

## Usage
Add `--kv-cache-dtype fp8_e5m2` to enable this feature. Currently it only works when flashinfer is not disabled.

## Performance & Accuracy
Tested with llama2-13b-chat on A100, the throughput increased by **17.8%** without accuracy degradation.
|  Enable fp8_e5m2 kv cache| Throughput | MMLU (nsub=10) Avg Accuracy | gsm8k Accuracy |
| --------------------- | ---------- | ------------------------------- | -------------- |
| N | 7.04 req/s | 0.487                           | 0.340          |
| Y  | 8.30 req/s | 0.488                           | 0.334          |

<details>
<summary>Reproduce</summary>

```
# w/o fp8_e5m2 kv cache
python3 -m sglang.launch_server --model-path meta-llama/Llama-2-13b-chat-hf --port 30000 --trust-remote-code --disable-radix-cache --tp=1
# w/ fp8_e5m2 kv cache
python3 -m sglang.launch_server --model-path meta-llama/Llama-2-13b-chat-hf --port 30000 --trust-remote-code --disable-radix-cache --tp=1 --kv-cache-dtype fp8_e5m2

# benchmark
python3 -m sglang.bench_serving --backend sglang --tokenizer meta-llama/Llama-2-13b-chat-hf  --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts 1000 --request-rate 128

# evaluation
python3 benchmark/mmlu/bench_sglang.py --nsub 10
python3 benchmark/gsm8k/bench_sglang.py --num-questions 1319
```
</details>

The performance boost is model dependent. llama3-8b was also tested, but the performance was not improved.

### CODE REVIEW

**Comment by zhyncs** [comment]
Nice work! I'll review it asap. May we also support FP8 E4M3?

**Comment by ispobock** [comment]
> May we also support FP8 E4M3?

FP8 E4M3 needs scale factor and calibration. We may add it in the future.

**Review by zhyncs — APPROVED** [review]

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 1]
```diff
@@ -112,6 +120,28 @@ def clear(self):
         # We also add one slot. This slot is used for writing dummy output from padded tokens.
         self.mem_state[0] = False
 
+    @abstractmethod
+    def get_key_buffer(self, layer_id: int) -> torch.Tensor:
+        raise NotImplementedError
```
Comment (merrymercy) [inline python/sglang/srt/mem_cache/memory_pool.py]:
```suggestion
        raise NotImplementedError()
```

**Review by merrymercy — APPROVED** [review]

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -334,6 +334,21 @@ def init_memory_pool(
         max_num_reqs: int = None,
         max_total_tokens: int = None,
     ):
+        if self.server_args.kv_cache_dtype == "auto":
+            self.kv_cache_dtype = self.dtype
+        elif self.server_args.kv_cache_dtype == "fp8_e5m2":
+            if self.server_args.disable_flashinfer or self.server_args.enable_mla:
```
Comment (zhyncs) [inline python/sglang/srt/model_executor/model_runner.py:340]: Currently, only FlashInfer is supported and not Triton, due to the issue of insufficient smem. This needs to be fixed in another PR.

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -123,26 +153,52 @@ def __init__(
...
+        loc: torch.Tensor,
+        cache_k: torch.Tensor,
+        cache_v: torch.Tensor,
+    ):
+        if cache_k.dtype != self.dtype:
+            cache_k = cache_k.to(self.dtype)
+        if cache_v.dtype != self.dtype:
+            cache_v = cache_v.to(self.dtype)
+        if self.store_dtype != self.dtype:
+            self.k_buffer[layer_id][loc] = cache_k.view(self.store_dtype)
```
Comment (zhyncs) [inline python/sglang/srt/mem_cache/memory_pool.py:197]: workaround for float8_e5m2
> Store as torch.uint8 because Tensor index_put is not implemented for torch.float8_e5m2

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 4]
```diff
@@ -112,6 +120,28 @@ def clear(self):
         # We also add one slot. This slot is used for writing dummy output from padded tokens.
         self.mem_state[0] = False
 
+    @abstractmethod
+    def get_key_buffer(self, layer_id: int) -> torch.Tensor:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_value_buffer(self, layer_id: int) -> torch.Tensor:
+        raise NotImplementedError
```
Comment (merrymercy) [inline python/sglang/srt/mem_cache/memory_pool.py]:
```suggestion
        raise NotImplementedError()
```

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 5]
```diff
@@ -112,6 +120,28 @@ def clear(self):
...
+
+    @abstractmethod
+    def get_kv_buffer(self, layer_id: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        raise NotImplementedError
```
Comment (merrymercy) [inline python/sglang/srt/mem_cache/memory_pool.py]:
```suggestion
        raise NotImplementedError()
```

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 6]
```diff
@@ -112,6 +120,28 @@ def clear(self):
...
+
+    @abstractmethod
+    def set_kv_buffer(
+        self,
+        layer_id: int,
+        loc: torch.Tensor,
+        cache_k: torch.Tensor,
+        cache_v: torch.Tensor,
+    ) -> None:
+        raise NotImplementedError
```
Comment (merrymercy) [inline python/sglang/srt/mem_cache/memory_pool.py]:
```suggestion
        raise NotImplementedError()
```

**Comment by qeternity** [comment]
Sorry to dig this up but - are we suggesting that fp8 kv cache *increased* accuracy in both mmlu and gsm8k? Are we sure we don't have those values in the table reversed?

**Comment by ispobock** [comment]
@qeternity In the previous evaluation, I tested gsm8k with 200 questions (default setting in the benchmark script), so the result may not be reliable enough. I just test all the datasets and update the result in the table.
